In [1]:
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm
import numpy as np
import pandas as pd
from bp import data_gather_from_files,run_strategy_optimised,run_strategy_eval
import time
from datetime import datetime
from dateutil.relativedelta import relativedelta

In [3]:
def error_check(params,n):
    if len(params) != n:
        raise IndexError('The number of parameters is not correct')

def generate_date_ranges_for_walk_forward(start_month_year, end_month_year, day=15, n_months = 1):
    # Initialize an empty list to store the date ranges
    date_ranges = []
    if day not in range(1, 29):
        raise ValueError('Day must be between 1 and 28')
    # Convert the input strings to datetime objects, using the given day
    start_date = datetime.strptime(f"{day} {start_month_year}", '%d %b %Y')
    end_date = datetime.strptime(f"{day} {end_month_year}", '%d %b %Y')
    
    # Generate the date ranges
    current_date = start_date
    while current_date <= end_date:
        next_date = current_date + relativedelta(months=n_months)
        date_range = [current_date.strftime('%d %b %Y'), (next_date - relativedelta(days=1)).strftime('%d %b %Y')]
        date_ranges.append(date_range)
        current_date = next_date
    
    return date_ranges[:-1]

def get_previous_n_months(end_date_str, n_months):
    # Convert the input string to a datetime object
    end_date = datetime.strptime(end_date_str, '%d %b %Y')
    
    # Calculate the start date
    start_date = end_date - relativedelta(months=n_months)
    
    # Create the date range
    date_range = [start_date.strftime('%d %b %Y'), (end_date - relativedelta(days=1)).strftime('%d %b %Y')]
    
    return date_range

In [22]:
def walk_forward_analysis(evaluation_start, evaluation_end, evaluation_day,parameters,optimization_function = None, optimizer_params =[],  lookback_in_months = 6,evaluation_period = 3):
    generated_date_ranges = generate_date_ranges_for_walk_forward(evaluation_start, evaluation_end,evaluation_day,n_months = evaluation_period)
    df = {}
    for dates in generated_date_ranges:
        train_period = get_previous_n_months(dates[0], lookback_in_months)
        train_data = data_gather_from_files(train_period[0],train_period[1])['EURUSD.mid']
        print('Data gathered for training period: ',train_period[0],train_period[1])
        test_data = data_gather_from_files(dates[0],dates[1])['EURUSD.mid']
        print('Data gathered for testing period: ',dates[0],dates[1])
        max_loss, R_PNL,profit,optimal_params = optimization_function(train_data,test_data,parameters,optimizer_params)
        print('Optimal parameters are: ',optimal_params)
        #add optimization phrase evaluation
        a, b, c = run_strategy_optimised(train_data, optimal_params[0],optimal_params[1])
        print('Optimization phrase; Max loss,R_PNL,profit are: ',a, b, c)
        ##
        print('Testing phrase; Max loss, R_PNL,profit are: ',max_loss,R_PNL,profit)
        df[dates[0] +'-'+ dates[1]] = [max_loss, R_PNL,profit]
    df = pd.DataFrame(df).T
    df.columns = ['max_loss','R_PNL','profit']
    return df

In [5]:
import warnings
warnings.simplefilter("ignore", category=RuntimeWarning)

## Group 1 optimization

In [6]:
import random
from deap import base, creator, tools

def deap_optimiser_g_n(train_data, test_data, parameters, optimization_params):
    ngen = optimization_params[0]  # number of generations
    npop = optimization_params[1]  # number of population

    error_check(parameters,2)
    
    grid_params = parameters[0]
    position_params = parameters[1]

    creator.create("FitnessMax", base.Fitness, weights=(1.0,)) #maximizing
    creator.create("Individual", list, fitness=creator.FitnessMax)

    toolbox = base.Toolbox()

    # Define the genes for our individual
    toolbox.register("G_gene", random.randint, grid_params[0]//grid_params[2], grid_params[1]//grid_params[2])
    toolbox.register("n_gene", random.randint, position_params[0]//position_params[2], position_params[1]//position_params[2])

    # Create an individual with the genes
    toolbox.register("individual", tools.initCycle, creator.Individual, (toolbox.G_gene, toolbox.n_gene), n=1)
    toolbox.register("population", tools.initRepeat, list, toolbox.individual)

    def objective(individual):
        G, n = individual[0]*grid_params[2], individual[1]*position_params[2]
        max_loss, R_PNL, profit = run_strategy_optimised(train_data, G, n)
    
        constraints = [
            max_loss < -500e3
        ]

        if any(constraints):
            return float('-inf'),  # Return large negative value when constraints are not satisfied
        return profit,

    toolbox.register("mate", tools.cxTwoPoint)
    toolbox.register("mutate", tools.mutUniformInt, low=[grid_params[0]//grid_params[2], position_params[0]//position_params[2]], 
                     up=[grid_params[1]//grid_params[2], position_params[1]//position_params[2]], indpb=0.2)
    toolbox.register("select", tools.selTournament, tournsize=3)   ##############
    toolbox.register("evaluate", objective)

    population = toolbox.population(n=npop)
    CXPB, MUTPB = 0.5, 0.2

    # Evaluate the entire population
    fitnesses = list(map(toolbox.evaluate, population))
    for ind, fit in zip(population, fitnesses):
        ind.fitness.values = fit

    #initiate early stopping
    stagnant_generations = 0  # Counter for generations without improvement
    MAX_STAGNANT_GEN = optimization_params[2]  # Early stopping criterion: stop if no improvement over x generations
    best_fitness_so_far = float('-inf')  # since we're maximizing
    ##
    for gen in range(ngen):
        offspring = toolbox.select(population, len(population))
        offspring = list(map(toolbox.clone, offspring))
        
        # Crossover
        for child1, child2 in zip(offspring[::2], offspring[1::2]):
            if random.random() < CXPB:
                toolbox.mate(child1, child2)
                del child1.fitness.values
                del child2.fitness.values
        
        # Mutation
        for mutant in offspring:
            if random.random() < MUTPB:
                toolbox.mutate(mutant)
                del mutant.fitness.values

        fitnesses = list(map(toolbox.evaluate, offspring))
        for ind, fit in zip(offspring, fitnesses):
            ind.fitness.values = fit

        #early stopping
        current_best_fitness = max(ind.fitness.values[0] for ind in population)

        if current_best_fitness > best_fitness_so_far:
            best_fitness_so_far = current_best_fitness
            stagnant_generations = 0  # Reset counter
        else:
            stagnant_generations += 1

        if stagnant_generations >= MAX_STAGNANT_GEN:
            print(f"Early stopping on generation {gen} due to no improvement.")
            break
        ##
        population[:] = offspring

    best_ind = tools.selBest(population, 1)[0]
    optimal_g = np.round(best_ind[0]*grid_params[2],5)
    optimal_n = best_ind[1]*position_params[2]
    print("optimisation completed")
    max_loss, R_PNL,profit = run_strategy_optimised(test_data, optimal_g,optimal_n)
    return max_loss, R_PNL,profit,[optimal_g,optimal_n]


In [9]:
grid_params = [0.001,0.01,0.0005]
lot_params = [100000,2000000,100000]

n_grid_params = ((grid_params[1]-grid_params[0])/grid_params[2])
n_lot_params = ((lot_params[1]-lot_params[0])/lot_params[2])

n_combination = n_grid_params*n_lot_params
print('number of grid params:-',(n_grid_params))
print('number of lot params:-',(n_lot_params))
print('total_number_of_combinations:-',(n_combination))

# Adjust these parameter according to search space
n_trials = 10 #NGEN
npop = 100
early_stopping_gen = 10 # no early stopping
optimizer_param = [n_trials, npop, early_stopping_gen]

parameters = [grid_params,lot_params]
results = walk_forward_analysis('jan 2021','jan 2022',1,parameters,optimization_function=deap_optimiser_g_n,optimizer_params=optimizer_param,lookback_in_months=6,evaluation_period=3)
results

number of grid params:- 18.0
number of lot params:- 19.0
total_number_of_combinations:- 342.0
Data gathered for training period:  01 Jul 2020 31 Dec 2020
Data gathered for testing period:  01 Jan 2021 31 Mar 2021
optimisation completed
Optimal parameters are:  [0.0025, 2000000]
Optimization phrase; Max loss,R_PNL,profit are:  -264108.8869 148947.5583 127750.6832
Training phrase; Max loss, R_PNL,profit are:  -41600.0 62500.0001 -41600.0
Data gathered for training period:  01 Oct 2020 31 Mar 2021
Data gathered for testing period:  01 Apr 2021 30 Jun 2021
optimisation completed
Optimal parameters are:  [0.0095, 1600000]
Optimization phrase; Max loss,R_PNL,profit are:  -76000.0 258400.0 258400.0
Training phrase; Max loss, R_PNL,profit are:  -121600.0 77320.0 98400.0
Data gathered for training period:  01 Jan 2021 30 Jun 2021
Data gathered for testing period:  01 Jul 2021 30 Sep 2021
optimisation completed
Optimal parameters are:  [0.0095, 2000000]
Optimization phrase; Max loss,R_PNL,profit

max_loss       R_PNL   profit
01 Jan 2021-31 Mar 2021  -41600.0  62500.0001 -41600.0
01 Apr 2021-30 Jun 2021 -121600.0  77320.0000  98400.0
01 Jul 2021-30 Sep 2021       0.0  68900.0000  61800.0
01 Oct 2021-31 Dec 2021  -28800.0  48888.8889  66240.0

## Group 1 Test run again for stability (Tune parameters)

In [10]:
grid_params = [0.001,0.01,0.0005]
lot_params = [100000,2000000,100000]

n_grid_params = ((grid_params[1]-grid_params[0])/grid_params[2])
n_lot_params = ((lot_params[1]-lot_params[0])/lot_params[2])

n_combination = n_grid_params*n_lot_params
print('number of grid params:-',(n_grid_params))
print('number of lot params:-',(n_lot_params))
print('total_number_of_combinations:-',(n_combination))

# Adjust these parameter according to search space
n_trials = 10 #NGEN
npop = 100
early_stopping_gen = 10 # no early stopping
optimizer_param = [n_trials, npop, early_stopping_gen]

parameters = [grid_params,lot_params]
results = walk_forward_analysis('jan 2021','jan 2022',1,parameters,optimization_function=deap_optimiser_g_n,optimizer_params=optimizer_param,lookback_in_months=6,evaluation_period=3)
results

number of grid params:- 18.0
number of lot params:- 19.0
total_number_of_combinations:- 342.0
Data gathered for training period:  01 Jul 2020 31 Dec 2020
Data gathered for testing period:  01 Jan 2021 31 Mar 2021
optimisation completed
Optimal parameters are:  [0.0025, 2000000]
Optimization phrase; Max loss,R_PNL,profit are:  -264108.8869 148947.5583 127750.6832
Training phrase; Max loss, R_PNL,profit are:  -41600.0 62500.0001 -41600.0
Data gathered for training period:  01 Oct 2020 31 Mar 2021
Data gathered for testing period:  01 Apr 2021 30 Jun 2021
optimisation completed
Optimal parameters are:  [0.009, 1300000]
Optimization phrase; Max loss,R_PNL,profit are:  -81900.0 245699.9999 245700.0
Training phrase; Max loss, R_PNL,profit are:  -93600.0 72306.0 91000.0
Data gathered for training period:  01 Jan 2021 30 Jun 2021
Data gathered for testing period:  01 Jul 2021 30 Sep 2021
optimisation completed
Optimal parameters are:  [0.0095, 2000000]
Optimization phrase; Max loss,R_PNL,profi

max_loss       R_PNL   profit
01 Jan 2021-31 Mar 2021  -41600.0  62500.0001 -41600.0
01 Apr 2021-30 Jun 2021  -93600.0  72306.0000  91000.0
01 Jul 2021-30 Sep 2021       0.0  68900.0000  61800.0
01 Oct 2021-31 Dec 2021  -28800.0  48888.8889  66240.0

### parameters seem stable with n_trials = 10 , npop = 100, no early stopping, run time = 13 mins

## Group 2 Optimization

In [21]:
import random
from deap import base, creator, tools

def deap_optimiser_multiplier(train_data, test_data, parameters, optimization_params):
    ngen = optimization_params[0]  # number of generations
    npop = optimization_params[1]  # number of population

    error_check(parameters,4)
    
    grid_params = parameters[0]
    position_params = parameters[1]
    multiplier_params = parameters[2]
    lookback_params = parameters[3]

    creator.create("FitnessMax", base.Fitness, weights=(1.0,)) #maximizing
    creator.create("Individual", list, fitness=creator.FitnessMax)

    toolbox = base.Toolbox()

    # Define the genes for our individual
    toolbox.register("G_gene", random.randint, grid_params[0]//grid_params[2], grid_params[1]//grid_params[2])
    toolbox.register("n_gene", random.randint, position_params[0]//position_params[2], position_params[1]//position_params[2])

    toolbox.register("multiplier_gene", random.randint, multiplier_params[0]//multiplier_params[2], multiplier_params[1]//multiplier_params[2])
    toolbox.register("lookback_gene", random.choice, lookback_params)

    CXPB, MUTPB = 0.5, 0.2

    def custom_mutate(individual): 
        if random.random() < MUTPB:
            individual[0] = random.randint(grid_params[0]//grid_params[2], grid_params[1]//grid_params[2])
        if random.random() < MUTPB:
            individual[1] = random.randint(position_params[0]//position_params[2], position_params[1]//position_params[2])
        if random.random() < MUTPB:
            individual[2] = random.randint(multiplier_params[0]//multiplier_params[2], multiplier_params[1]//multiplier_params[2])
        if random.random() < MUTPB:
            individual[3] = random.choice(lookback_params)
        return individual,

    toolbox.register("mutate", custom_mutate)

    # Create an individual with the genes
    toolbox.register("individual", tools.initCycle, creator.Individual, (toolbox.G_gene, toolbox.n_gene, toolbox.multiplier_gene, toolbox.lookback_gene), n=1)
    toolbox.register("population", tools.initRepeat, list, toolbox.individual)

    def objective(individual):
        G, n, multiplier, lookback = individual[0]*grid_params[2], individual[1]*position_params[2], individual[2]*multiplier_params[2], individual[3]

        max_loss, R_PNL, profit = run_strategy_optimised(train_data, G, n, multiplier = multiplier,lookback = lookback)
    
        constraints = [
            max_loss < -500e3
        ]

        if any(constraints):
            return float('-inf'),  # Return large negative value when constraints are not satisfied
        return profit,

    toolbox.register("mate", tools.cxTwoPoint)
    toolbox.register("select", tools.selTournament, tournsize=3)   
    toolbox.register("evaluate", objective)

    population = toolbox.population(n=npop)

    # Evaluate the entire population
    fitnesses = list(map(toolbox.evaluate, population))
    for ind, fit in zip(population, fitnesses):
        ind.fitness.values = fit

    #initiate early stopping
    stagnant_generations = 0  # Counter for generations without improvement
    MAX_STAGNANT_GEN = optimization_params[2]  # Early stopping criterion: stop if no improvement over x generations
    best_fitness_so_far = float('-inf')  # since we're maximizing
    ##
    for gen in range(ngen):
        offspring = toolbox.select(population, len(population))
        offspring = list(map(toolbox.clone, offspring))
        
        # Crossover
        for child1, child2 in zip(offspring[::2], offspring[1::2]):
            if random.random() < CXPB:
                toolbox.mate(child1, child2)
                del child1.fitness.values
                del child2.fitness.values
        
        # Mutation
        for mutant in offspring:
            if random.random() < MUTPB:
                toolbox.mutate(mutant)
                del mutant.fitness.values

        fitnesses = list(map(toolbox.evaluate, offspring))
        for ind, fit in zip(offspring, fitnesses):
            ind.fitness.values = fit

        #early stopping
        current_best_fitness = max(ind.fitness.values[0] for ind in population)

        if current_best_fitness > best_fitness_so_far:
            best_fitness_so_far = current_best_fitness
            stagnant_generations = 0  # Reset counter
        else:
            stagnant_generations += 1

        if stagnant_generations >= MAX_STAGNANT_GEN:
            print(f"Early stopping on generation {gen} due to no improvement.")
            break
        ##
        population[:] = offspring

    best_ind = tools.selBest(population, 1)[0]
    optimal_g = np.round(best_ind[0]*grid_params[2],5)
    optimal_n = best_ind[1]*position_params[2]
    optimal_m = np.round(best_ind[2] *multiplier_params[2],5)
    optimal_l = best_ind[3]

    print("optimisation completed")
    max_loss, R_PNL,profit = run_strategy_optimised(test_data, optimal_g,optimal_n,multiplier = optimal_m,lookback = optimal_l)
    return max_loss, R_PNL,profit,[optimal_g,optimal_n,optimal_m,optimal_l]


In [29]:
def walk_forward_analysis_multiplier(evaluation_start, evaluation_end, evaluation_day,parameters,optimization_function = None, optimizer_params =[],  lookback_in_months = 6,evaluation_period = 3):
    generated_date_ranges = generate_date_ranges_for_walk_forward(evaluation_start, evaluation_end,evaluation_day,n_months = evaluation_period)
    df = {}
    for dates in generated_date_ranges:
        train_period = get_previous_n_months(dates[0], lookback_in_months)
        train_data = data_gather_from_files(train_period[0],train_period[1])['EURUSD.mid']
        print('Data gathered for training period: ',train_period[0],train_period[1])
        test_data = data_gather_from_files(dates[0],dates[1])['EURUSD.mid']
        print('Data gathered for testing period: ',dates[0],dates[1])
        max_loss, R_PNL,profit,optimal_params = optimization_function(train_data,test_data,parameters,optimizer_params)
        print('Optimal parameters are: ',optimal_params)
        #add optimization phrase evaluation
        a, b, c = run_strategy_optimised(train_data, optimal_params[0],optimal_params[1], multiplier = optimal_params[2], lookback=optimal_params[3])
        print('Optimization phrase; Max loss,R_PNL,profit are: ',a, b, c)
        ##
        print('Testing phrase; Max loss, R_PNL,profit are: ',max_loss,R_PNL,profit)
        df[dates[0] +'-'+ dates[1]] = [max_loss, R_PNL,profit]
    df = pd.DataFrame(df).T
    df.columns = ['max_loss','R_PNL','profit']
    return df

In [36]:
grid_params = [0.001,0.01,0.0005]
lot_params = [100000,2000000,100000]
multiplier_params = [1,3,0.1]
lookback_params = [9,20,30,50,100,150,200]

n_grid_params = ((grid_params[1]-grid_params[0])/grid_params[2])
n_lot_params = ((lot_params[1]-lot_params[0])/lot_params[2])
n_multiplier_params = ((multiplier_params[1]-multiplier_params[0])/multiplier_params[2])
n_lookback_params = len(lookback_params)

n_combination = n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params

print('number of grid params:-',(n_grid_params))
print('number of lot params:-',(n_lot_params))
print('number of multiplier params:-',(n_multiplier_params))
print('number of lookback params:-',(n_lookback_params))
print('total_number_of_combinations:-',(n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params))


# Adjust these parameter according to search space
n_trials = 2 #NGEN
npop = 5
early_stopping_gen = 10 # early stopping
optimizer_param = [n_trials, npop, early_stopping_gen]

parameters = [grid_params,lot_params,multiplier_params,lookback_params]
results = walk_forward_analysis_multiplier('jan 2021','jan 2022',1,parameters,optimization_function=deap_optimiser_multiplier,optimizer_params=optimizer_param,lookback_in_months=6,evaluation_period=3)
results

number of grid params:- 18.0
number of lot params:- 19.0
number of multiplier params:- 20.0
number of lookback params:- 7
total_number_of_combinations:- 47880.0
Data gathered for training period:  01 Jul 2020 31 Dec 2020
Data gathered for testing period:  01 Jan 2021 31 Mar 2021
optimisation completed
Optimal parameters are:  [0.0025, 1800000, 2.3, 200]
Optimization phrase; Max loss,R_PNL,profit are:  -53955.44 242771.7571 248171.5631
Testing phrase; Max loss, R_PNL,profit are:  -13500.0 144091.4503 110979.909
Data gathered for training period:  01 Oct 2020 31 Mar 2021
Data gathered for testing period:  01 Apr 2021 30 Jun 2021
optimisation completed
Optimal parameters are:  [0.0065, 1500000, 2.6, 9]
Optimization phrase; Max loss,R_PNL,profit are:  0 418470.0 409602.0
Testing phrase; Max loss, R_PNL,profit are:  -214687.5 -204937.5 -190537.5
Data gathered for training period:  01 Jan 2021 30 Jun 2021
Data gathered for testing period:  01 Jul 2021 30 Sep 2021
optimisation completed
Optim

max_loss        R_PNL      profit
01 Jan 2021-31 Mar 2021  -13500.0  144091.4503  110979.909
01 Apr 2021-30 Jun 2021 -214687.5 -204937.5000 -190537.500
01 Jul 2021-30 Sep 2021       0.0   48230.0000   43260.000
01 Oct 2021-31 Dec 2021  -25200.0   92653.0464   96452.000

In [31]:
data = data_gather_from_files('1 jul 2020','31 dec 2020')['EURUSD.mid']
run_strategy_optimised(data, 0.004, 2000000, multiplier=1.9, lookback=100)

(-112880.0, 154152.0, 150466.0)

In [33]:
grid_params = [0.001,0.01,0.0005]
lot_params = [100000,2000000,100000]
multiplier_params = [1,3,0.1]
lookback_params = [9,20,30,50,100,150,200]

n_grid_params = ((grid_params[1]-grid_params[0])/grid_params[2])
n_lot_params = ((lot_params[1]-lot_params[0])/lot_params[2])
n_multiplier_params = ((multiplier_params[1]-multiplier_params[0])/multiplier_params[2])
n_lookback_params = len(lookback_params)

n_combination = n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params

print('number of grid params:-',(n_grid_params))
print('number of lot params:-',(n_lot_params))
print('number of multiplier params:-',(n_multiplier_params))
print('number of lookback params:-',(n_lookback_params))
print('total_number_of_combinations:-',(n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params))


# Adjust these parameter according to search space
n_trials = 10 #NGEN
npop = 100
early_stopping_gen = 10 # early stopping
optimizer_param = [n_trials, npop, early_stopping_gen]

parameters = [grid_params,lot_params,multiplier_params,lookback_params]
results = walk_forward_analysis_multiplier('jan 2021','jan 2022',1,parameters,optimization_function=deap_optimiser_multiplier,optimizer_params=optimizer_param,lookback_in_months=6,evaluation_period=3)
results

number of grid params:- 18.0
number of lot params:- 19.0
number of multiplier params:- 20.0
number of lookback params:- 7
total_number_of_combinations:- 47880.0
Data gathered for training period:  01 Jul 2020 31 Dec 2020
Data gathered for testing period:  01 Jan 2021 31 Mar 2021
optimisation completed
Optimal parameters are:  [0.0055, 1000000, 2.4, 100]
Optimization phrase; Max loss,R_PNL,profit are:  0 781521.1721 877340.0
Testing phrase; Max loss, R_PNL,profit are:  -16720.0 81699.9999 81700.0
Data gathered for training period:  01 Oct 2020 31 Mar 2021
Data gathered for testing period:  01 Apr 2021 30 Jun 2021
optimisation completed
Optimal parameters are:  [0.0055, 1200000, 2.6, 20]
Optimization phrase; Max loss,R_PNL,profit are:  -3960.0 529560.0001 540816.0
Testing phrase; Max loss, R_PNL,profit are:  -242152.9412 -235552.9412 -228832.9412
Data gathered for training period:  01 Jan 2021 30 Jun 2021
Data gathered for testing period:  01 Jul 2021 30 Sep 2021
optimisation completed
O

max_loss        R_PNL       profit
01 Jan 2021-31 Mar 2021  -16720.0000   81699.9999   81700.0000
01 Apr 2021-30 Jun 2021 -242152.9412 -235552.9412 -228832.9412
01 Jul 2021-30 Sep 2021  -11466.0000   78032.8421   12885.6000
01 Oct 2021-31 Dec 2021  -22625.0000  174595.0857  158225.0000

## Revised walk forward to return both table

In [72]:
def walk_forward_analysis_multiplier(evaluation_start, evaluation_end, evaluation_day, parameters, optimization_function=None, optimizer_params=[], lookback_in_months=6, evaluation_period=3):
    generated_date_ranges = generate_date_ranges_for_walk_forward(evaluation_start, evaluation_end, evaluation_day, n_months=evaluation_period)
    #replace df here
    testing_results = {} # use to be df
    optimization_results = {}
    
    for dates in generated_date_ranges:
        train_period = get_previous_n_months(dates[0], lookback_in_months)
        train_data = data_gather_from_files(train_period[0], train_period[1])['EURUSD.mid']
        print('Data gathered for training period:', train_period[0], train_period[1])
        test_data = data_gather_from_files(dates[0], dates[1])['EURUSD.mid']
        print('Data gathered for testing period:', dates[0], dates[1])
        max_loss, R_PNL, profit, optimal_params = optimization_function(train_data, test_data, parameters, optimizer_params)
        print('Optimal parameters are:', optimal_params)
        
        # Optimization phase evaluation
        opt_max_loss, opt_R_PNL, opt_profit = run_strategy_optimised(train_data, optimal_params[0], optimal_params[1], multiplier=optimal_params[2], lookback=optimal_params[3])
        print('Optimization phase; Max loss, R_PNL, profit are:', opt_max_loss, opt_R_PNL, opt_profit)
        optimization_results[train_period[0] + '-' + train_period[1]] = [opt_max_loss, opt_R_PNL, opt_profit]
        
        # Testing phase results
        print('Testing phase; Max loss, R_PNL, profit are:', max_loss, R_PNL, profit)
        testing_results[dates[0] + '-' + dates[1]] = [max_loss, R_PNL, profit]

    testing_df = pd.DataFrame(testing_results).T
    testing_df.columns = ['max_loss', 'R_PNL', 'profit']

    optimization_df = pd.DataFrame(optimization_results).T
    optimization_df.columns = ['max_loss', 'R_PNL', 'profit']
    
    return testing_df, optimization_df


In [74]:
grid_params = [0.001,0.01,0.0005]
lot_params = [100000,2000000,100000]
multiplier_params = [1,3,0.1]
lookback_params = [9,20,30,50,100,150,200]

n_grid_params = ((grid_params[1]-grid_params[0])/grid_params[2])
n_lot_params = ((lot_params[1]-lot_params[0])/lot_params[2])
n_multiplier_params = ((multiplier_params[1]-multiplier_params[0])/multiplier_params[2])
n_lookback_params = len(lookback_params)

n_combination = n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params

print('number of grid params:-',(n_grid_params))
print('number of lot params:-',(n_lot_params))
print('number of multiplier params:-',(n_multiplier_params))
print('number of lookback params:-',(n_lookback_params))
print('total_number_of_combinations:-',(n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params))


# Adjust these parameter according to search space
n_trials = 50 #NGEN
npop = 100
early_stopping_gen = n_trials # no early stopping
optimizer_param = [n_trials, npop, early_stopping_gen]

parameters = [grid_params,lot_params,multiplier_params,lookback_params]

testing_df, optimization_df = walk_forward_analysis_multiplier('jan 2021','jan 2022',1,parameters,optimization_function=deap_optimiser_multiplier,optimizer_params=optimizer_param,lookback_in_months=6,evaluation_period=3)


number of grid params:- 18.0
number of lot params:- 19.0
number of multiplier params:- 20.0
number of lookback params:- 7
total_number_of_combinations:- 47880.0
Data gathered for training period: 01 Jul 2020 31 Dec 2020
Data gathered for testing period: 01 Jan 2021 31 Mar 2021
optimisation completed
Optimal parameters are: [0.006, 1400000, 2.5, 30]
Optimization phase; Max loss, R_PNL, profit are: 0 768599.2146 866740.0
Testing phase; Max loss, R_PNL, profit are: -77420.0 2645.4545 -77420.0
Data gathered for training period: 01 Oct 2020 31 Mar 2021
Data gathered for testing period: 01 Apr 2021 30 Jun 2021
optimisation completed
Optimal parameters are: [0.0025, 100000, 2.5, 20]
Optimization phase; Max loss, R_PNL, profit are: -12523.4375 474642.0129 565822.8125
Testing phase; Max loss, R_PNL, profit are: -21056.7436 5122.1721 32240.1314
Data gathered for training period: 01 Jan 2021 30 Jun 2021
Data gathered for testing period: 01 Jul 2021 30 Sep 2021
optimisation completed
Optimal param

In [75]:
optimization_df

max_loss        R_PNL       profit
01 Jul 2020-31 Dec 2020      0.0000  768599.2146  866740.0000
01 Oct 2020-31 Mar 2021 -12523.4375  474642.0129  565822.8125
01 Jan 2021-30 Jun 2021  -4500.0000  457075.1963  465515.0000
01 Apr 2021-30 Sep 2021  -1760.0000  382122.3998  382122.4000

In [76]:
testing_df

max_loss        R_PNL       profit
01 Jan 2021-31 Mar 2021 -77420.0000    2645.4545  -77420.0000
01 Apr 2021-30 Jun 2021 -21056.7436    5122.1721   32240.1314
01 Jul 2021-30 Sep 2021 -16245.0000   71701.3923    9617.0000
01 Oct 2021-31 Dec 2021 -49984.0000  166320.0000  177091.2000

In [81]:
grid_params = [0.001,0.01,0.0005]
lot_params = [100000,2000000,100000]
multiplier_params = [1,3,0.1]
lookback_params = [9,20,30,50,100,150,200]

n_grid_params = ((grid_params[1]-grid_params[0])/grid_params[2])
n_lot_params = ((lot_params[1]-lot_params[0])/lot_params[2])
n_multiplier_params = ((multiplier_params[1]-multiplier_params[0])/multiplier_params[2])
n_lookback_params = len(lookback_params)

n_combination = n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params

print('number of grid params:-',(n_grid_params))
print('number of lot params:-',(n_lot_params))
print('number of multiplier params:-',(n_multiplier_params))
print('number of lookback params:-',(n_lookback_params))
print('total_number_of_combinations:-',(n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params))


# Adjust these parameter according to search space
n_trials = 50 #NGEN
npop = 100
early_stopping_gen = n_trials # no early stopping
optimizer_param = [n_trials, npop, early_stopping_gen]

parameters = [grid_params,lot_params,multiplier_params,lookback_params]

testing_df, optimization_df = walk_forward_analysis_multiplier('jan 2021','jan 2022',1,parameters,optimization_function=deap_optimiser_multiplier,optimizer_params=optimizer_param,lookback_in_months=6,evaluation_period=3)


number of grid params:- 18.0
number of lot params:- 19.0
number of multiplier params:- 20.0
number of lookback params:- 7
total_number_of_combinations:- 47880.0
Data gathered for training period: 01 Jul 2020 31 Dec 2020
Data gathered for testing period: 01 Jan 2021 31 Mar 2021
optimisation completed
Optimal parameters are: [0.006, 1100000, 2.4, 200]
Optimization phase; Max loss, R_PNL, profit are: 0 712270.7209 826465.2
Testing phase; Max loss, R_PNL, profit are: -154000.88 -4045.358 -154000.88
Data gathered for training period: 01 Oct 2020 31 Mar 2021
Data gathered for testing period: 01 Apr 2021 30 Jun 2021
optimisation completed
Optimal parameters are: [0.0035, 1500000, 2.5, 20]
Optimization phase; Max loss, R_PNL, profit are: -34125.0 578812.5002 575737.5
Testing phase; Max loss, R_PNL, profit are: -236999.9999 -214249.9999 -205499.9999
Data gathered for training period: 01 Jan 2021 30 Jun 2021
Data gathered for testing period: 01 Jul 2021 30 Sep 2021
optimisation completed
Optimal

In [82]:
optimization_df

max_loss        R_PNL      profit
01 Jul 2020-31 Dec 2020       0.0  712270.7209  826465.200
01 Oct 2020-31 Mar 2021  -34125.0  578812.5002  575737.500
01 Jan 2021-30 Jun 2021  -20244.0  494476.9350  495796.000
01 Apr 2021-30 Sep 2021    -504.0  383209.5228  458876.428

In [83]:
testing_df

max_loss        R_PNL       profit
01 Jan 2021-31 Mar 2021 -154000.8800   -4045.3580 -154000.8800
01 Apr 2021-30 Jun 2021 -236999.9999 -214249.9999 -205499.9999
01 Jul 2021-30 Sep 2021 -212940.0000 -110224.7995 -110224.8000
01 Oct 2021-31 Dec 2021  -26883.8400  156404.3831  156404.3822

### experiment

In [77]:
grid_params = [0.001,0.01,0.0005]
lot_params = [100000,2000000,100000]
multiplier_params = [1,3,0.1]
lookback_params = [9,20,30,50,100,150,200]

n_grid_params = ((grid_params[1]-grid_params[0])/grid_params[2])
n_lot_params = ((lot_params[1]-lot_params[0])/lot_params[2])
n_multiplier_params = ((multiplier_params[1]-multiplier_params[0])/multiplier_params[2])
n_lookback_params = len(lookback_params)

n_combination = n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params

print('number of grid params:-',(n_grid_params))
print('number of lot params:-',(n_lot_params))
print('number of multiplier params:-',(n_multiplier_params))
print('number of lookback params:-',(n_lookback_params))
print('total_number_of_combinations:-',(n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params))


# Adjust these parameter according to search space
n_trials = 30 #NGEN
npop = 100
early_stopping_gen = n_trials # no early stopping
optimizer_param = [n_trials, npop, early_stopping_gen]

parameters = [grid_params,lot_params,multiplier_params,lookback_params]

testing_df, optimization_df = walk_forward_analysis_multiplier('jan 2021','jan 2022',1,parameters,optimization_function=deap_optimiser_multiplier,optimizer_params=optimizer_param,lookback_in_months=6,evaluation_period=3)


number of grid params:- 18.0
number of lot params:- 19.0
number of multiplier params:- 20.0
number of lookback params:- 7
total_number_of_combinations:- 47880.0
Data gathered for training period: 01 Jul 2020 31 Dec 2020
Data gathered for testing period: 01 Jan 2021 31 Mar 2021
optimisation completed
Optimal parameters are: [0.006, 1100000, 2.4, 100]
Optimization phase; Max loss, R_PNL, profit are: 0 712270.7209 826465.2
Testing phase; Max loss, R_PNL, profit are: -154000.88 -4045.358 -154000.88
Data gathered for training period: 01 Oct 2020 31 Mar 2021
Data gathered for testing period: 01 Apr 2021 30 Jun 2021
optimisation completed
Optimal parameters are: [0.0035, 1400000, 2.6, 20]
Optimization phase; Max loss, R_PNL, profit are: -40180.0 599760.0 599592.0
Testing phase; Max loss, R_PNL, profit are: -242550.7409 -220745.7409 -211190.7409
Data gathered for training period: 01 Jan 2021 30 Jun 2021
Data gathered for testing period: 01 Jul 2021 30 Sep 2021
optimisation completed
Optimal pa

In [78]:
optimization_df

max_loss        R_PNL      profit
01 Jul 2020-31 Dec 2020       0.0  712270.7209  826465.200
01 Oct 2020-31 Mar 2021  -40180.0  599760.0000  599592.000
01 Jan 2021-30 Jun 2021   -8250.0  423036.3984  440000.000
01 Apr 2021-30 Sep 2021  -11100.0  330944.8041  443835.504

In [79]:
testing_df

max_loss        R_PNL       profit
01 Jan 2021-31 Mar 2021 -154000.8800   -4045.3580 -154000.8800
01 Apr 2021-30 Jun 2021 -242550.7409 -220745.7409 -211190.7409
01 Jul 2021-30 Sep 2021  -79062.5000    -357.5000    -357.5000
01 Oct 2021-31 Dec 2021 -190445.9999 -110541.9785 -131204.3999

### Train:Test 3:1 ratio

In [84]:
grid_params = [0.001,0.01,0.0005]
lot_params = [100000,2000000,100000]
multiplier_params = [1,3,0.1]
lookback_params = [9,20,30,50,100,150,200]

n_grid_params = ((grid_params[1]-grid_params[0])/grid_params[2])
n_lot_params = ((lot_params[1]-lot_params[0])/lot_params[2])
n_multiplier_params = ((multiplier_params[1]-multiplier_params[0])/multiplier_params[2])
n_lookback_params = len(lookback_params)

n_combination = n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params

print('number of grid params:-',(n_grid_params))
print('number of lot params:-',(n_lot_params))
print('number of multiplier params:-',(n_multiplier_params))
print('number of lookback params:-',(n_lookback_params))
print('total_number_of_combinations:-',(n_grid_params*n_lot_params*n_multiplier_params*n_lookback_params))


# Adjust these parameter according to search space
n_trials = 50 #NGEN
npop = 100
early_stopping_gen = n_trials # no early stopping
optimizer_param = [n_trials, npop, early_stopping_gen]

parameters = [grid_params,lot_params,multiplier_params,lookback_params]

testing_df, optimization_df = walk_forward_analysis_multiplier('jan 2021','jan 2022',1,parameters,optimization_function=deap_optimiser_multiplier,optimizer_params=optimizer_param,lookback_in_months=3,evaluation_period=1)


number of grid params:- 18.0
number of lot params:- 19.0
number of multiplier params:- 20.0
number of lookback params:- 7
total_number_of_combinations:- 47880.0
Data gathered for training period: 01 Oct 2020 31 Dec 2020
Data gathered for testing period: 01 Jan 2021 31 Jan 2021
optimisation completed
Optimal parameters are: [0.0065, 500000, 2.9, 150]
Optimization phase; Max loss, R_PNL, profit are: 0 339269.3955 565075.75
Testing phase; Max loss, R_PNL, profit are: -20540.0 7537.0 7537.0
Data gathered for training period: 01 Nov 2020 31 Jan 2021
Data gathered for testing period: 01 Feb 2021 28 Feb 2021
optimisation completed
Optimal parameters are: [0.0015, 500000, 2.2, 200]
Optimization phase; Max loss, R_PNL, profit are: -19176.0 515132.7983 505508.0
Testing phase; Max loss, R_PNL, profit are: -118506.0 -84146.2304 -51964.64
Data gathered for training period: 01 Dec 2020 28 Feb 2021
Data gathered for testing period: 01 Mar 2021 31 Mar 2021
optimisation completed
Optimal parameters are

In [85]:
optimization_df

max_loss        R_PNL       profit
01 Oct 2020-31 Dec 2020       0.0  339269.3955  565075.7500
01 Nov 2020-31 Jan 2021  -19176.0  515132.7983  505508.0000
01 Dec 2020-28 Feb 2021 -114000.0  320921.8750  320381.5625
01 Jan 2021-31 Mar 2021    -400.0  123709.9241  384144.4270
01 Feb 2021-30 Apr 2021  -26325.0  345281.9402  351266.5000
01 Mar 2021-31 May 2021  -19200.0  292679.4610  291104.0084
01 Apr 2021-30 Jun 2021  -22819.5  290369.3645  301570.5000
01 May 2021-31 Jul 2021  -38400.0  239660.3202  291900.0000
01 Jun 2021-31 Aug 2021  -19890.0  299578.5001  294582.6000
01 Jul 2021-30 Sep 2021       0.0  153600.0000  256588.8000
01 Aug 2021-31 Oct 2021   -2890.0  227677.1288  298125.6000
01 Sep 2021-30 Nov 2021   -1890.0  319881.9924  416661.3600

In [86]:
testing_df

max_loss        R_PNL      profit
01 Jan 2021-31 Jan 2021  -20540.000    7537.0000    7537.000
01 Feb 2021-28 Feb 2021 -118506.000  -84146.2304  -51964.640
01 Mar 2021-31 Mar 2021 -138474.375  -50646.8750 -138474.375
01 Apr 2021-30 Apr 2021  -10780.000   -5345.1474  -10780.000
01 May 2021-31 May 2021  -90480.000  -76348.3726  -85636.200
01 Jun 2021-30 Jun 2021  -16639.992  141695.9360  159743.894
01 Jul 2021-31 Jul 2021  -87087.000  -67529.0002  -65824.000
01 Aug 2021-31 Aug 2021  -14400.000    5360.0000    5360.000
01 Sep 2021-30 Sep 2021  -79601.600   11700.0000  -79601.600
01 Oct 2021-31 Oct 2021       0.000    6720.0000    6720.000
01 Nov 2021-30 Nov 2021  -80920.000   -3715.7143  -24310.000
01 Dec 2021-31 Dec 2021   -8378.100   69849.0000   92684.520

# Summary

342 combs
ngen=10 npop=100, 13 mins >>nice

47880 combs 

Test with walkforward 6-3

* ngen= 50 npop = 100, 56 mins >> walk forward can be good and bad

* ngen= 30 npop = 100, 34 mins >> opt okay but walk forward really bad

Test with walkforward 3-1

* ngen= 50 npop = 100,  mins >> walk forward